### PR TITLE
dia: 2017-06-22 → 2020-04-07

### PR DESCRIPTION
**Pull request by jtojnar**
###### Motivation for this change
Getting rid of GNOME 2 package set https://github.com/NixOS/nixpkgs/issues/39976

cc @viric

###### Things done

- [ x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by veprbl** [comment]
This fails with
```
2020-04-07 04:35:09.971 dia[70685:3631021] *** WARNING: Method userSpaceScaleFactor in class NSView is deprecated on 10.7 and later. It should not be used in new applications. Use convertRectToBacking: instead. 

** (dia:70685): WARNING **: 04:35:10.471: Unable to find object type: GRAFCET - Step

** (dia:70685): WARNING **: 04:35:10.471: Unable to find object type: GRAFCET - Step

<many similar warnings>

** (dia:70685): WARNING **: 04:35:10.564: Unable to find object type: Flowchart - Predefined Process

** (dia:70685): WARNING **: 04:35:10.564: Unable to find object type: Standard - Box

** (dia:70685): CRITICAL **: 04:35:10.574: Couldn't find standard objects when looking for object-libs in '/nix/store/jbwri07nqs0qcybyv576rfr12id4n08v-dia-unstable-2020-04-07/lib/dia/'; exiting...
```

I came up with a following workaround:
```diff
diff --git a/pkgs/applications/graphics/dia/default.nix b/pkgs/applications/graphics/dia/default.nix
index 6eee16145a7..5acba6f5895 100644
--- a/pkgs/applications/graphics/dia/default.nix
+++ b/pkgs/applications/graphics/dia/default.nix
@@ -58,6 +58,12 @@ stdenv.mkDerivation {
       build-aux/post-install.py
   '';
 
+  postInstall = stdenv.lib.optionalString stdenv.isDarwin ''
+    for dylib in "$out"/lib/dia/*; do
+      ln "''${dylib}" "''${dylib%.*}.so"
+    done
+  '';
+
   meta = with stdenv.lib; {
     description = "GNOME Diagram drawing software";
     homepage = "https://wiki.gnome.org/Apps/Dia";
```

After that dia launches it is more or less usable, except for some rendering problems:
dia

**Comment by jtojnar** [comment]
@veprbl pushed a new revision, re-adding gtk-mac-integration-gtk2 and merging https://gitlab.gnome.org/GNOME/dia/-/merge_requests/68. Does that fix it for you?

**Comment by veprbl** [comment]
@jtojnar Thank you for implementing the fix. It does what needs to be done, but misses some of the libraries:
```
$ ls -1 /nix/store/jaq7jss0hmlglpsjb4w7bjgp9g7y5rpx-dia-unstable-2020-04-07/lib/dia/
libaadl_objects.dylib
libcairo_filter.so
libcgm_filter.so
libchronogram_objects.dylib
libcustom_lines_objects.dylib
libcustom_objects.dylib
libdb_objects.dylib
libdxf_filter.so
liber_objects.dylib
libflowchart_objects.dylib
libfs_objects.dylib
libgrafcet_objects.dylib
libhpgl_filter.so
libistar_objects.dylib
libjackson_objects.dylib
libkaos_objects.dylib
liblayout_filter.so
libmetapost_filter.so
libmisc_objects.dylib
libnetwork_objects.dylib
libpdf_filter.so
libpgf_filter.so
libpixbuf_filter.so
libpostscript_filter.so
libpstricks_filter.so
libpython_plugin.so
libsadt_objects.dylib
libshape_filter.so
libstandard_objects.dylib
libsvg_filter.so
libuml_objects.dylib
libvdx_filter.so
libwpg_filter.so
libxfig_filter.so
libxslt_filter.dylib
```

Regarding the rendering issues, I don't see much change. I've played with a window a bit more, I've noticed it blanks the window on resize, but depending on the size of the window three different things can happen. If the window is "smaller", it will render window as black (which macOS seems to screenshot as transparent color):

dia black

If it is "bigger" it actually renders the proper grey background and able to render most of the icons:

dia grey

And, finally, excessively small windows will be black, but will turn to white garbled texture on mouse hover (see my first screenshot for example).

Most buts get redrawn on mouse hover, so this is almost usable.

File: pkgs/applications/graphics/dia/default.nix [hunk 1]
```diff
@@ -1,37 +1,69 @@
...
-    ++ (with perlPackages; [ perl XMLParser ]);
+  buildInputs = [
+    glib
+    gtk2
+    libxml2
+    python2
+    zlib
+    cairo
+  ] ++ stdenv.lib.optionals stdenv.isDarwin [
+    gtk-mac-integration-gtk2
```
Comment (veprbl) [inline pkgs/applications/graphics/dia/default.nix:55]: Doesn't seem like this is used:
```
# grep -ri gtk_mac src/
# grep -ri gtk-mac src/
#
```

Comment (jtojnar) [reply pkgs/applications/graphics/dia/default.nix:55]: I would expect it to be somehow magically linked. Is it not listed in `nix-store --query --tree` or `nix why-depends`?

Comment (veprbl) [reply pkgs/applications/graphics/dia/default.nix:55]:
```
# nix-store --query --tree /nix/store/qflajk99al4hk1893gz6z2m996x6l930-dia-0.97.3.20170622 | grep mac
+---/nix/store/brvfal9n3wi562r98q1rgicxcjlyh7i1-gtk-mac-integration-2.0.8
|   +---/nix/store/brvfal9n3wi562r98q1rgicxcjlyh7i1-gtk-mac-integration-2.0.8 [...]
# nix-store --query --tree /nix/store/jaq7jss0hmlglpsjb4w7bjgp9g7y5rpx-dia-unstable-2020-04-07 | grep mac
#
```

Comment (jtojnar) [reply pkgs/applications/graphics/dia/default.nix:55]: Hmm, I can still grep `GtkosxApplication` in the source code. Looks like `HAVE_MAC_INTEGRATION` was not ported to Meson.

**Comment by veprbl** [comment]
Just tested 2889df2a66e, it starts up fine.

File: pkgs/applications/graphics/dia/default.nix (hunk 1 above)
Comment (veprbl) [reply pkgs/applications/graphics/dia/default.nix:64]: Strangely enough, old dia also doesn't have the bug:

Screenshot 2020-04-07 at 17 01 29

And it does use the system menubar, like it's supposed to do with gtk-mac-integration.

Comment (jtojnar) [reply pkgs/applications/graphics/dia/default.nix:64]: Does the latest change help?

**Comment by ZanderBrown** [comment]
Just checking: Your shipping this as unstable/testing/devel right?

File: pkgs/applications/graphics/dia/default.nix (hunk 1 above)
Comment (veprbl) [reply pkgs/applications/graphics/dia/default.nix:64]: @jtojnar The build is now failing with
```
../app/interface.c:1060:34: error: use of undeclared identifier 'dia_app_icon'
        gdk_pixbuf_new_from_inline (-1, dia_app_icon, FALSE, NULL));
                                        ^
1 error generated.
```
After applying
```
diff --git a/app/interface.c b/app/interface.c
--- a/app/interface.c
+++ b/app/interface.c
@@ -1057,7 +1057,7 @@ _create_mac_integration (GtkWidget *menubar)
     gtk_widget_hide (menubar); /* not working, it's shown elsewhere */
     /* setup the dock icon */
     gtkosx_application_set_dock_icon_pixbuf (theOsxApp,
-       gdk_pixbuf_new_from_inline (-1, dia_app_icon, FALSE, NULL));
+       pixbuf_from_resource ("/org/gnome/Dia/icons/org.gnome.Dia.png"));
   }
   /* Don't quit without asking to save files first */
   g_signal_connect (theOsxApp, "NSApplicationBlockTermination",
```
I get it to compile and run. I now get a proper dia icon in the dock and system-wide menu works again. There is no improvement with rendering.

**Comment by jtojnar** [comment]
@ZanderBrown Yup, we use `dia-unstable` as the package name.

**Comment by jtojnar** [comment]
@veprbl Would you be willing to bisect the graphical glitches? Should take roughly eight steps.

(You would run `git bisect start master b86085dfe2b048a2d37d587adf8ceba6fb8bc43c` in the Dia repo, change the nixpkgs expression’s src to your dia checkout, and then build it and try to run it and mark the bisection step with `git bisect good/bad`.)

**Comment by veprbl** [comment]
@jtojnar 
The bisect didn't go as smooth as planned. Many builds failed for various reasons and there was some bug that caused dia to crash on start. I patched couple of the bugs on the go and arrived to the following expression:
```
with import <nixpkgs> {};

stdenv.mkDerivation {
  name = "dia";

  src = lib.cleanSource ./diagit;

  nativeBuildInputs = [
    meson
    ninja
    pkg-config
    gettext
    python3 # for install script
    gtk2 # for gtk-update-icon-cache
    #libxslt # for xsltproc
    desktop-file-utils
    intltool
  ];

  buildInputs = [
    glib
    gtk2
    libxml2
    python2
    zlib
    cairo
  ];

  postPatch = ''
    patchShebangs \
      generate_run_with_dia_env.sh \
      build-aux/post-install.py
    substituteInPlace plug-ins/wmf/wmf.cpp \
      --replace '#include "config.h"' ""
    substituteInPlace objects/standard/outline.c \
      --replace '#include "tool-icons.h"' ""
  '';

  postInstall = ''
    for dylib in "$out"/lib/dia/*.dylib; do
      ln "''${dylib}" "''${dylib%.*}.so"
    done
  '';
}
```
What I learned was that there is a "good" version of dia that is able to be built with this (or a similar) meson-based expression, which is close enough to what you propose in this PR.

```
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
079712821232fcdd19f593e01a8855d331c8d833
a31108180eeb20c7a2bd05beb007b200d53b0113
5fcae38699fcc363915d723b79f480288942e48a
6830485cee2f3194cc62aa7678e38b7ceb130000
ae8bc6cd85aa2014d2d041fe8dc1adb4824315cd
f8dbd900db7633cfd6d20293c559e40f7d0a3196
7ef24ee68d0f40125bf8069493a939ea9641511e
b3e047388ac8736586533d726ac78061313af517
292266f5e9bea534c2cc0288e4312d429aae881c
f1de9111d1dc5bc175de229d4aebe1a32cb79198
03ac6beac229289964da7fcae5674ad56fb4fe59
53fef431b2b51192850a958d5cbdd01e4fe0ce2a
1bef8c839986ec6ca89a44061154c483e3a19281
5f17286651785e6fee902c7d908c8420e918fe00
3dc03f2617175cd6f1ee6bc936f1afcf8a1afc5b
5654328926b2b4872b4e6056c3bf51adf47b0fa2
223d893f993bd4d4b4c93e82cd04020d255cfec7
33e5aaa3585f040898159bc16d1d7ca67523b0de
9566e0b188c826a9de04765215a458501f6b1c8e
d0137e84d8cb47feb873911fd49ae18ff180c9bc
af3c7836c0bf9aa7ac62fbac9a3dd569a9c3db81
d7fd3c783920fdd3ab2f0aca1796b3c9a507978d
216ba0d4882035ba309e8ba4b670fbeb413c9365
659179878502e515b6748c3e76dd3d51422a1925
665f9429dc7ad9a27cc6c0c265e586f37a5db33a
dda471808eba97bcb0d3cbc79e78e15134cde4b6
8d7fc4f4a2a54f91faecda065894a14fedbe1afb
ea21bcab9d8740cbf7feef81359525128734cea6
ddc1c7070e93aeaa16703e1764c96b07599d3dae
fd5d59244fd92909896c640578a6fca940aef47f
2dbbb055c4aa190148b23dacec976259a1ddcccc
651f7eb4810de4f94f04a106b0a45eb0495b1381
642282d842de4e59a06c0d7b53418683352a6a65
a1e4b53edc31ee705f32836cb60e89f34c316367
c1ec976f1e450fc5cc9873b929db0bf3b96b90ae
e99aff67206ed75466f8607307b65b770adf04d4
2e864255fdfa9c08d9e1897d660f9fc60ba72a73
ff4c247dea72c436e0c3b206039eb7e72dd60c6c
1b7726cd9044852af07fa296e2722237ef1878ee
7637543cb20d35e18ae1b824f28e3fc3a11016c4
04758526c0c42f76f0978f1408aea4c526ab6673
9927ab22db91508f594f1d31f365cd02631e6c28
81d1ad7fb78027465120c228b85f74b5df502841
d0339fa23a511feb44e653655dd24b191bf6b899
54e780ca98b43f35b74a5e4aba86f961b2e717bb
5e148bac88bded7d9e21f34b2f30f24ae0c1e9eb
24c914354f87c08e2653779e560cc3292b211c62
dc563ee9401ad4ee03ca80683502216e9b84401e
0a0ec28e8d886d63c5689ab8af8dab0f4e43d7ff
cbee573bffcf942a9bb77559e388996ee58fad9c
5c65e84c7d3bfe078cfe7b34a511b26bcb8f30d9
ec665e5ad60519ad7d93c378cbcbd11edf092f69
4ede4a82866ed4745d87bfadc99669b3b14e4f1d
ba44b6db7813ab7bcb56a89e70139fcba52ebc79
69c0d45c50c3bda8388e4a044c06114f6c75dacb
70880b81b445eb016ebde8bf95f534cdea9e7af8
5607703367db4edf32900f6c982054c13f0fc6d5
8fafd1ac49417ae99f25bcf3d81e50517aa12f79
84b1a30dc99e1521c9bfda4516166d6d3ec0b9a2
ae01344346f0286e587593669c8124994948ca11
8d8ea9e6bef88bd4c943856aa57a81745e86bf5a
c27cf10e59f9d5990972e8bab5be5c238f537f9c
91d93cfa61049bac47d00914f476c10413916ab4
eca725c5cca9262c98ad059cfeb904cbbbd5b84e
5370b223102038cc4e022631486ac02ab9ac9901
efe7a0f4dea2de2ab31b4bab642db457866ced23
40d710e82eb524f2f23ecd97435cbb51130e4d2e
f318b78a7e1ca98d46af8f743a6bdcdeb0d41a7b
dab6d2e7ce8154eda1b9b6eb1fa215abc5065c18
db4101d90eb69d2bfe0ca55d598d1167b0719b9a
154e1ba0ecc7da1ed85466c0a056cd43453679f5
7f291ba554e8dbcfced20639b11e3efe46aa3630
c9ed7166259ec383d909ba0989b8ffe05aff8f3a
d940918bc768eec62d0918063bc1e5a1bbace89e
778dd62b8fc02dd182452f8fd30390d28e4f59da
c83cf0ec3770936f754208a4631fd0c5be465479
26e50b4f63df135b0da6cf11abccb155897a37c7
892bb8a648af1843fbe8b3fc0b367cfb9398e350
55c8e61fe1cc4d5a6e86973e9a96633c0092fdeb
We cannot bisect more!
```

**Comment by jtojnar** [comment]
Yeah, unfortunately, it is not easy to automate it. At best you will be able to keep `automake` & co. and `preConfigure` from the original expression and use the following hack to disable meson when not available:

```nix
  ${if builtins.pathExists "${src}/meson.build" then null else "dontUseMesonConfigure"} = true;
  ${if builtins.pathExists "${src}/meson.build" then null else "dontUseNinjaBuild"} = true;
  ${if builtins.pathExists "${src}/meson.build" then null else "dontUseNinjaInstall"} = true;
 ```

With 8 steps, it is also feasible, if annoying, to run the bisection manually.

**Comment by veprbl** [comment]
@jtojnar I think you got me wrong. The breakage happened after meson was introduced. I am now trying to bisect the crash to learn how to fix it and, maybe, narrow down the search.

**Comment by veprbl** [comment]
I guess, this has to do with new cairo renderer in dia.

**Comment by ZanderBrown** [comment]
Most likely, interestingly Windows & X11 haven't had an issue like that with rendering the canvas

However as of a few days ago line/arrow chooser in the properties dialogue do, except that code hasn't changed in months

It seems when rendering goes wrong in gtk2 it goes really wrong

**Comment by jtojnar** [comment]
The question is do we ship this, or wait for a fix (possibly GTK 3 port)?

**Comment by veprbl** [comment]
@jtojnar I would suggest we pin dia to a latest pre-cairo version for darwin. I can prepare a change for that.

**Comment by ZanderBrown** [comment]
This exact version? I'd say no, even as an unstable package that's way too broken

Of course I'm interested what's going on, especially as it's almost certainly my fault. Might try a VM at somepoint

**Comment by veprbl** [comment]
@ZanderBrown Let me know if there is anything I can try to help debugging this.

**Comment by jtojnar** [comment]
Okay, since it does not really improve that much and it is far from the worst gnome2 library usage offender :crying_cat_face:, we will table this PR for now.

**Comment by ZanderBrown** [comment]
> far from the worst gnome2 library usage offender

What is it using? if it's libgnomeui there is a configure flag for that

**Comment by jtojnar** [comment]
`libart_lgpl` and `libgnomeui`

**Comment by ZanderBrown** [comment]
Ah well until we figure out the cairo issue libart_lgpl is sticking around, for libgnomeui just `--disable` it, isn't needed

**Comment by armeenm** [comment]
Any updates on this?

**Comment by veprbl** [comment]
Superseded by #209587

**Pull request closed without merging**